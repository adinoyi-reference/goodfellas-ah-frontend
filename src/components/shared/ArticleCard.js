import React from 'react';
import likesIcon from '../../assets/icons8-heart-outline-48-grey.png';
import authorProfileImage from '../../assets/john.jpg';

const Card = ({ articles }) => {
  const CardList = articles.map((eachCard) => {
    const splicedBody = eachCard.body.slice(0, 150);
    return (
      <div className="row col-lg-3 hero-card" key={eachCard.id}>
        <div className="hero-card-details col-sm-7">
          <h6>{eachCard.title}</h6>
          <p>
            {splicedBody}
          </p>
          <div className="row hero-card-author">
            <img
              className="authorProfile"
              src={authorProfileImage}
              alt="Author Profile"
            />
            <p className="col-lg-9">{eachCard.user.firstname} {eachCard.user.lastname}</p>
          </div>
        </div>
        <div className="hero-card-image col-sm-5">
          <img src={likesIcon} alt="" />
          <p>1300</p>
        </div>
      </div>
    );
  });
  return (
    <div className="row card-wrapper">
      { CardList }
    </div>
  );
};

export default Card;
