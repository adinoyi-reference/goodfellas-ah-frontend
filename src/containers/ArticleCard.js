import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import parser from 'react-html-parser';
import { getArticles } from '../actions/articleActions';
import Loading from '../components/shared/Loading';
import { filterReactions } from '../mixin';

export class Card extends Component {
  state = {
    articleLimit: 6,
  };

  componentDidMount() {
    const { getArticles: getAllArticles } = this.props;
    getAllArticles();
  }

  sortArticlesByLikes = (articles) => {
    const sortedArticles = articles
      .map((eachArticle) => {
        const filteredArticles = filterReactions(eachArticle.reactions).likes;
        return { article: eachArticle, track: filteredArticles };
      })
      .sort((a, b) => {
        const result = b.track - a.track;
        return result;
      });
    return sortedArticles;
  };

  displayArticles = (articleLimit) => {
    const { articles } = this.props;
    const newlySortedArticles = this.sortArticlesByLikes(articles);
    const displayedArticles = newlySortedArticles.slice(0, 0 + articleLimit);
    return displayedArticles;
  };

  handleClick = (e) => {
    const { articleLimit } = this.state;
    if (articleLimit === 18) {
      e.currentTarget.innerHTML = '';
    }
    if (articleLimit !== 24) {
      this.setState({ articleLimit: articleLimit + 6 });
      this.displayCards(this.displayArticles);
    }
  };


  changeBackground = (image) => {
    const changeBackground = {
      backgroundImage: `linear-gradient(rgba(0, 0, 0, .32), rgba(0, 0, 0, .32)), url(${image}`,
      backgroundSize: 'cover',
    };
    return changeBackground;
  };

  displayCards = (displayArticles) => {
    const { articleLimit } = this.state;
    return displayArticles(articleLimit).map((card) => {
      const displayedDescription = card.article.description.slice(0, 120);
      const displayedTitle = card.article.title.slice(0, 24);
      return (
        <Link key={card.article.id} to={`/articles/${card.article.slug}`}>
          <div
            onClick={this.getArticle}
            className="hero-card"
          >
            <div className="hero-card-details col-sm-7">
              <h6>{parser(displayedTitle)}</h6>
              {parser(displayedDescription)}
              <div className="hero-card-author">
                {card.article.user.profile.image === null ? (
                  <img
                    src="https://res.cloudinary.com/drmmqcxkc/image/upload/v1541581893/Authors%20Haven/user-placeholder.png"
                    alt="Author Profile"
                  />
                ) : (
                  <img
                    src={card.article.user.profile.image}
                    alt="Author Profile"
                  />
                )}
                <p>
                  {card.article.user.firstname} {card.article.user.lastname}
                </p>
              </div>
            </div>
            {card.article.image === null ? (
              <div className="hero-card-image col-sm-5">
                <img
                  src="https://res.cloudinary.com/drmmqcxkc/image/upload/v1541426068/Authors%20Haven/icons8-heart-outline-48-grey.png"
                  alt=""
                />
                <p>{filterReactions(card.article.reactions).likes}</p>
              </div>
            ) : (
              <div
                style={this.changeBackground(card.article.image)}
                className="hero-card-image col-sm-5"
              >
                <img
                  src="https://res.cloudinary.com/drmmqcxkc/image/upload/v1541426068/Authors%20Haven/icons8-heart-outline-48-grey.png"
                  alt=""
                />
                <p>{filterReactions(card.article.reactions).likes}</p>
              </div>
            )}
          </div>
        </Link>
      );
    });
  };

  render() {
    const { articles } = this.props;
    return (
      <div>
        {articles ? (
          <div className="card-wrapper">
            {this.displayCards(this.displayArticles)}
            <div>
              {articles.length <= 6 ? (
                null
              ) : (
                <div onClick={this.handleClick} className="hero-moreArticles row">
                  <p ref="moreArticles">More Articles</p>
                  <img
                    src="https://res.cloudinary.com/drmmqcxkc/image/upload/v1541426068/Authors%20Haven/icons8-expand-arrow-24.png"
                    alt=""
                  />
                </div>
              )}
            </div>
          </div>
        ) : (
          <Loading />
        )}
      </div>
    );
  }
}

function mapStatetoProps(state) {
  return { articles: state.articles.articles.articles, error: state.articles.error.response };
}

export default connect(mapStatetoProps, { getArticles })(Card);
