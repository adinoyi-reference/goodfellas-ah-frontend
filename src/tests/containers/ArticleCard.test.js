import React from 'react';
import { mount } from 'enzyme';
import { MemoryRouter } from 'react-router-dom';
import toJson from 'enzyme-to-json';
import ArticleCard from '../../containers/ArticleCard';
import Root from '../../root';

let wrapped;

beforeEach(() => {
  wrapped = mount(
    <Root>
      <MemoryRouter initialEntries={[{ key: 'testkey' }]}>
        <ArticleCard />
      </MemoryRouter>
    </Root>,
  );
});

// afterEach(() => wrapped.unmount());

describe('Article Card UI', () => {
  describe('render features', () => {
    test('component should render as expected', () => {
      const tree = toJson(wrapped);
      expect(tree).toMatchSnapshot();
    });
  });
});
