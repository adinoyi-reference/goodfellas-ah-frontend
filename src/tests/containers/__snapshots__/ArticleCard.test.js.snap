// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Article Card UI render features component should render as expected 1`] = `
<_default>
  <Provider
    store={
      Object {
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      }
    }
  >
    <MemoryRouter
      initialEntries={
        Array [
          Object {
            "key": "testkey",
          },
        ]
      }
    >
      <Router
        history={
          Object {
            "action": "POP",
            "block": [Function],
            "canGo": [Function],
            "createHref": [Function],
            "entries": Array [
              Object {
                "hash": "",
                "key": "testkey",
                "pathname": "/",
                "search": "",
              },
            ],
            "go": [Function],
            "goBack": [Function],
            "goForward": [Function],
            "index": 0,
            "length": 1,
            "listen": [Function],
            "location": Object {
              "hash": "",
              "key": "testkey",
              "pathname": "/",
              "search": "",
            },
            "push": [Function],
            "replace": [Function],
          }
        }
      >
        <Connect(Card)>
          <Card
            getArticles={[Function]}
          >
            <div>
              <Loading>
                <div
                  className="loading-icon"
                >
                  <img
                    alt=""
                    src="https://res.cloudinary.com/drmmqcxkc/image/upload/v1541426069/Authors%20Haven/Spinner-1s.gif"
                  />
                </div>
              </Loading>
              <div
                className="hero-moreArticles row"
                onClick={[Function]}
              >
                <p>
                  More Articles
                </p>
                <img
                  alt=""
                  src="https://res.cloudinary.com/drmmqcxkc/image/upload/v1541426068/Authors%20Haven/icons8-expand-arrow-24.png"
                />
              </div>
            </div>
          </Card>
        </Connect(Card)>
      </Router>
    </MemoryRouter>
  </Provider>
</_default>
`;
